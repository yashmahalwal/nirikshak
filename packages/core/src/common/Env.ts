/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
