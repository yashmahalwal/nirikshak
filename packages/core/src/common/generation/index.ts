export { generateFaker } from "./fakerGen";
export { generateCustom } from "./customGen";
