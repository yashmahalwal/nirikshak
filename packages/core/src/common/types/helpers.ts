import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
