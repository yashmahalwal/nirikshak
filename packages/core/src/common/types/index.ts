export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
