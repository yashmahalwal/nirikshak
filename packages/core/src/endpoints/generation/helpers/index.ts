export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
