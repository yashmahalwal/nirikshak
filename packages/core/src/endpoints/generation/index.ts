export { generateURL } from "./urlStringGen";
export { getFromResource } from "./resourceStringGen";
export * from "./helpers";
export * from "./input";
