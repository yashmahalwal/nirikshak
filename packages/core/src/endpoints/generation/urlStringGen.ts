import { URLString, normalizeURLString } from "../types/urlString";
import { ResourceInstance } from "../../resource/types/helper";
import { Literal, isLiteral } from "../../common/types/literals";
import { isFakerString } from "../../common/types/fakerTypes";
import { generateFaker } from "../../common/generation/fakerGen";
import { isCustomFunctionString } from "../../common/types/custom";
import { SchemaHelpers } from "../../common/types/helpers";
import { generateCustom } from "../../common/generation/customGen";
import { getFromResource } from "./resourceStringGen";
import { ResourceString } from "../types/resourceString";

export async function generateURL(
    urlString: URLString,
    instance: ResourceInstance,
    helpers: SchemaHelpers
): Promise<string> {
    const components = normalizeURLString(urlString).split("/");
    const urlArr: Literal[] = [];
    for (const component of components) {
        if (component[0] === "{" && component[component.length - 1] === "}") {
            // Faker or custom or resource
            const slice = component.slice(1, component.length - 1);
            if (isFakerString(slice)) urlArr.push(generateFaker(slice));
            else if (isCustomFunctionString(slice))
                urlArr.push(await generateCustom(slice, helpers));
            else {
                const entry = getFromResource(
                    slice as ResourceString,
                    instance
                );
                if (!isLiteral(entry))
                    throw new Error(
                        `Invalid resource interpolation for url: ${slice}`
                    );
                urlArr.push(entry);
            }
        } else urlArr.push(component);
    }

    return urlArr.join("/");
}
