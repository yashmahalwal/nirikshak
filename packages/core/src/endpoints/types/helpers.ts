import { ResourceString, isResourceString } from "./resourceString";
import { FakerType, isFakerType } from "../../common/types/fakerTypes";
import {
    CustomFunctionType,
    isCustomFunction,
} from "../../common/types/custom";
import { Literal, isLiteral } from "../../common/types/literals";
import { Primitives } from "../../common/types/helpers";

export type MethodType = "DELETE" | "GET" | "PUT" | "POST" | "PATCH";
export function isMethodType(key: any): key is MethodType {
    if (typeof key !== "string") return false;

    switch (key) {
        case "DELETE":
        case "GET":
        case "PUT":
        case "POST":
        case "PATCH":
            return true;
        default:
            return false;
    }
}

export type Cases = "NEGATIVE" | "POSITIVE" | "DESTRUCTIVE";

export function isCaseType(key: any): key is Cases {
    if (typeof key !== "string") return false;
    switch (key) {
        case "NEGATIVE":
        case "POSITIVE":
        case "DESTRUCTIVE":
            return true;
        default:
            return false;
    }
}

export type BaseType =
    | FakerType
    | CustomFunctionType
    | ResourceString
    | Literal;

export function isBaseType(input: any): input is BaseType {
    return (
        isFakerType(input) ||
        isCustomFunction(input) ||
        isResourceString(input) ||
        isLiteral(input)
    );
}

export type Base = BaseType | Base[];

export function isBase(input: any): input is Base {
    if (Array.isArray(input)) return input.every((value) => isBase(value));
    return isBaseType(input);
}

export type WithModifiers<T extends Base | BodyType> = (T extends Base
    ? { type: Base }
    : { field: BodyType }) & {
    plural?: boolean;
    nullable?: boolean;
    optional?: boolean;
};

// Checking the validity of the modifier fields of the HOT
function checkModifiers(input: WithModifiers<Base | BodyType>): boolean {
    let value = true;
    void (
        "nullable" in input &&
        (value = value && typeof input["nullable"] === "boolean")
    );
    void (
        "optional" in input &&
        (value = value && typeof input["optional"] === "boolean")
    );
    void (
        "plural" in input &&
        (value = value && typeof input["plural"] === "boolean")
    );

    return value;
}

// type guard: WithModifiers<ResourceBase>
export function isWithModifiersBodyType(
    input: any
): input is WithModifiers<BodyType> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("field" in input) {
        return isBodyType(input["field"]);
    }
    return false;
}

// type guard: WithModifies<BaseType>
export function isWithModifiersBase(input: any): input is WithModifiers<Base> {
    if (!input || typeof input != "object") return false;

    if (!checkModifiers(input)) return false;

    if ("type" in input) {
        return isBase(input["type"]);
    }
    return false;
}

// type guard: WithModifiers<BaseType> or WithModifiers<ResourceBase>
// Infers from input
export function isWithModifiers(
    input: any
): input is WithModifiers<Base | BodyType> {
    return isWithModifiersBase(input) || isWithModifiersBodyType(input);
}

export type OneOfEntries =
    | { fields: (BodyType | WithModifiers<BodyType>)[] }
    | { types: (Base | WithModifiers<Base>)[] }
    | {
          fields: (BodyType | WithModifiers<BodyType>)[];
          types: (Base | WithModifiers<Base>)[];
      };

// type guard: OneOfEntries
export function isOneOfEntries(input: any): input is OneOfEntries {
    if (!input || typeof input !== "object") return false;
    if (!Array.isArray(input["fields"]) && !Array.isArray(input["types"]))
        return false;

    if (Array.isArray(input["fields"]) && !input["fields"].length) return false;
    if (Array.isArray(input["types"]) && !input["types"].length) return false;

    let value = true;
    value =
        value &&
        (Array.isArray(input["types"])
            ? input["types"].every(
                  (entry) => isBase(entry) || isWithModifiersBase(entry)
              )
            : true);

    value =
        value &&
        (Array.isArray(input["fields"])
            ? input["fields"].every(
                  (entry) => isBodyType(entry) || isWithModifiersBodyType(entry)
              )
            : true);

    return value;
}

export interface BodyType {
    [key: string]:
        | ResourceString
        | Exclude<Base, ResourceString>
        | WithModifiers<Base>
        | WithModifiers<BodyType>
        | OneOfEntries
        | BodyType;
}

export interface BodyInstance {
    [key: string]: Primitives | BodyInstance | BodyInstance[];
}

export function isBodyType(input: any): input is BodyType {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (entry) =>
            isResourceString(entry) ||
            isBase(entry) ||
            isWithModifiers(entry) ||
            isOneOfEntries(entry) ||
            isBodyType(entry)
    );
}

export interface HeaderMap {
    [key: string]: Exclude<BaseType, null>;
}

export function isHeaderMap(input: any): input is HeaderMap {
    if (!input || typeof input !== "object" || Array.isArray(input))
        return false;

    return Object.values(input).every(
        (value) => value !== null && isBaseType(value)
    );
}
