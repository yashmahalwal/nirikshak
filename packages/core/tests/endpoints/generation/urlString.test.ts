import faker from "faker";
import { URLString } from "../../../src/endpoints/types/urlString";
import { generateURL } from "../../../src/endpoints/generation/urlStringGen";
import { SchemaHelpers } from "../../../src/common/types/helpers";
import { RANDOMNESS_ITERATIONS } from "../../../src/common/Env";

const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidCases: {
    input: URLString;
    output: () => Promise<string>;
}[] = [
    {
        input: "home/classrooms/places",
        output: async (): Promise<string> => "/home/classrooms/places",
    },
    {
        input: "/resource/{faker:random.word}",
        output: async (): Promise<string> => "/resource/" + faker.random.word(),
    },
    {
        input: "/resource/{faker:random.word}/words/{custom:word}",
        output: async (): Promise<string> =>
            `/resource/${faker.random.word()}/words/${await Helpers.word()}`,
    },
    {
        input:
            "/resource/{resource:name.firstName}/words/{resource:name.lastName}",
        output: async (): Promise<string> =>
            `/resource/${resource.name.firstName}/words/${resource.name.lastName}`,
    },
    {
        input: `{resource:id}/numbers/{faker:random.number}/words/{custom:word}/zipCode/{resource:addresses[0].zipCode}/number/{resource:recursion.bases.args.plethora[3][0][2][0]}`,
        output: async (): Promise<string> =>
            `/${
                resource.id
            }/numbers/${faker.random.number()}/words/${await Helpers.word()}/zipCode/${
                resource.addresses[0].zipCode
            }/number/${resource.recursion.bases.args.plethora[3][0][2][0]}`,
    },
];

const InvalidInputs: string[] = [
    "resource/{resouce:lorem.ipsum}",
    "resource/{custom:lithium}",
    "resource/{custom:invalidHelper}",
    "resource/{resource:recursion.bases.args.plethora[3][0][2]}",
];

describe("URL strings", () => {
    
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
    ValidCases.forEach((entry, index) =>
        test(`Valid entry: ${index}, iteration: ${i}`, async () => {
            faker.seed(123 + index * i);
            const a = await generateURL(entry.input, resource, Helpers);
            faker.seed(123 + index * i);
            const b = await entry.output();
            expect(a).toEqual(b);
        })
    );

    test.each(InvalidInputs)(`Invalid entry: %#`, (entry) =>
        expect(generateURL(entry, resource, Helpers)).rejects.toMatchSnapshot()
    );
});
