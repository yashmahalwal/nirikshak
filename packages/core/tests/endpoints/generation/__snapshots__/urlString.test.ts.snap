// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URL strings Invalid entry: 0 1`] = `[Error: Invalid resource instance path: orem.ipsum]`;

exports[`URL strings Invalid entry: 1 1`] = `[Error: function: custom:lithium, args: ]`;

exports[`URL strings Invalid entry: 2 1`] = `[Error: Non literal type value from custom:invalidHelper]`;

exports[`URL strings Invalid entry: 3 1`] = `[Error: Invalid resource interpolation for url: resource:recursion.bases.args.plethora[3][0][2]]`;
